queue/listener.go

```diff
@@ -0,0 +1,94 @@
+package queue
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/formancehq/go-libs/v3/logging"
+)
+
+type (
+	CallbackFn func(ctx context.Context, metadata map[string]string, msg []byte) error
+)
+
+//go:generate mockgen -source listener.go -destination listener_generated.go -package queue . Listener
+type Listener interface {
+	Listen(ctx context.Context, ch <-chan *message.Message, fn CallbackFn) error
+	Done() <-chan struct{}
+}
+
+type listener struct {
+	wg *sync.WaitGroup
+
+	logger      logging.Logger
+	workerCount int
+	callbackFn  CallbackFn
+}
+
+func NewListener(
+	logger logging.Logger,
+	workerCount int,
+) Listener {
+	return &listener{
+		wg:          &sync.WaitGroup{},
+		logger:      logger,
+		workerCount: workerCount,
+	}
+}
+
+// Start polling SQS for messages
+func (l *listener) Listen(ctx context.Context, ch <-chan *message.Message, fn CallbackFn) error {
+	l.logger.WithField("workerCount", l.workerCount).Debugf("queue listener starting listen...")
+	l.callbackFn = fn
+	if l.workerCount < 1 {
```

Comment:
Why not check the constraints in the constructor since you can early detect the invalid configuration?